backend/packages/Upgrade/src/api/services/MoocletRewardsService.ts

```diff
@@ -0,0 +1,252 @@
+
+    return this.metricService.saveAllMetrics([metric], [context], logger);
+  }
+
+  public async parseLogsAndSendPotentialRewards(
+    user: RequestedExperimentUser,
+    logs: ILogInput[],
+    logger: UpgradeLogger
+  ): Promise<void> {
+    // First see if any simple metrics have valid reward values
+    const validSimpleMetricAttributes = this.gatherValidRewardMetrics(logs);
+
+    if (!validSimpleMetricAttributes.length) {
+      // no need to log anything here, just quit early as this is the most common case
+      return;
+    }
+
+    try {
+      // If we have valid metric rewards, only then check for active mooclet experiment refs
+      const moocletExperimentRefs = await this.getAllActiveMoocletExperimentRefs(logger);
+
+      if (!moocletExperimentRefs.length) {
+        logger.warn({
+          message: 'Reward metrics were logged, but no active mooclet experiment refs found.',
+          user,
+        });
+        return;
+      }
+
+      // Find each experiment match by reward metric key
+      const experimentsMatchingLoggedRewardKey = this.findExperimentByRewardMetricKey(
+        moocletExperimentRefs,
+        validSimpleMetricAttributes
+      );
+
+      if (!experimentsMatchingLoggedRewardKey.length) {
+        logger.warn({
+          message: 'Reward metrics were logged, but no active experiments matched the rewardMetricKeys',
+          moocletExperimentRefs,
+          validSimpleMetricAttributes,
+        });
+        return;
+      }
+
+      // For each match, find the user's condition, map to version id, and create reward object
+      experimentsMatchingLoggedRewardKey.forEach(async (experimentRewardPair) => {
+        const { moocletExperimentRef, rewardMetricValue } = experimentRewardPair;
+        const enrollment = await this.findEnrolledCondition(user.id, moocletExperimentRef.experimentId, logger);
+
+        // not sure if this even possible in reality, but should be handled
+        if (!enrollment) {
+          logger.error({
+            message: 'Could not find user enrollment for experiment.',
+            user,
+            moocletExperimentRef,
+          });
+          return;
+        }
+
+        const versionId = this.getVersionIdByConditionId(enrollment, moocletExperimentRef, logger);
+
+        if (!versionId) {
+          logger.error({
+            message: 'Could not find version id for user enrollment.',
+            enrollment,
+            moocletExperimentRef,
+          });
+          return;
+        }
+
+        if (versionId) {
+          const reward = {
+            variable: moocletExperimentRef.outcomeVariableName,
+            value: BinaryRewardMetricValueMap[rewardMetricValue],
+            mooclet: moocletExperimentRef.moocletId,
+            version: versionId,
+          };
+          logger.info({
+            message: 'Sending reward to mooclet',
+            reward,
+            user,
+          });
+          this.moocletDataService.postNewReward(reward, logger);
```

Comment:
`await` should be used, and if there is a way to send all the rewards in a single call, that would be better.

Reply:
it looks like I possibly _can_ send more than one reward at a time, I dug into their API and the django framework allows this if I send append `/create_many`, aka: `/value/create_many` and when I test this, I surprisingly can send multiple values in an array.

This isn't documented by them though, so I'm not 100% sure this will work as intended, would need to test out sending multiple. For MVP, I'd like to just send them one at a time as I've tested this quite a bit, and the limited usage that is planned for experiments with Mooclets at this point is that 1 `/log` will equal 1 reward at a time, because they will be firing a log close to moment they have completed the task that requires a reward, not in batches.